File: sdk/apps/ethereum/src/yielding_vaults.rs
use crate::{
    proto::definitions::{NetworkInfo, TokenInfo},
    tokens::unknown_token,
};

#[cfg(not(test))]
use alloc::{string::String, vec};
#[cfg(test)]
use std::{string::String, vec};

#[derive(PartialEq)]
pub(crate) struct VaultInfo {
    pub address: Option<[u8; 20]>,
    pub chain_id: Option<u64>,
    pub name: String,
    pub asset_token: TokenInfo,
    pub vault_token: TokenInfo,
}

pub(crate) fn unknown_vault() -> VaultInfo {
    VaultInfo {
        address: None,
        chain_id: None,
        name: "UNKNOWN VAULT".into(),
        asset_token: unknown_token(),
        vault_token: unknown_token(),
    }
}

fn known_vaults() -> [VaultInfo; 3] {
    [
        VaultInfo {
            // 0xa511d618cD0F9d7cAD791009d7c5E3b19c9568da
            address: Some([
                0xa5, 0x11, 0xd6, 0x18, 0xcd, 0x0f, 0x9d, 0x7c, 0xad, 0x79, 0x10, 0x09, 0xd7, 0xc5,
                0xe3, 0xb1, 0x9c, 0x95, 0x68, 0xda,
            ]),
            chain_id: Some(1),
            name: "Test Steakhouse USDC Prime Vault".into(),
            asset_token: TokenInfo {
                symbol: "USDC".into(),
                decimals: 6,
                address: vec![
                    0xa0, 0xb8, 0x69, 0x91, 0xc6, 0x21, 0x8b, 0x36, 0xc1, 0xd1, 0x9d, 0x4a, 0x2e,
                    0x9e, 0xb0, 0xce, 0x36, 0x06, 0xeb, 0x48,
                ],
                chain_id: 1,
                name: "USD Coin".into(),
            },
            vault_token: TokenInfo {
                symbol: "tstSHUSDCp".into(),
                decimals: 18,
                address: vec![
                    0xa5, 0x11, 0xd6, 0x18, 0xcd, 0x0f, 0x9d, 0x7c, 0xad, 0x79, 0x10, 0x09, 0xd7,
                    0xc5, 0xe3, 0xb1, 0x9c, 0x95, 0x68, 0xda,
                ],
                chain_id: 1,
                name: "Test Steakhouse USDC Prime Vault".into(),
            },
        },
        VaultInfo {
            // 0xde6c23E561F3e55846207EC45A91b777e0F7C889
            address: Some([
                0xde, 0x6c, 0x23, 0xe5, 0x61, 0xf3, 0xe5, 0x58, 0x46, 0x20, 0x7e, 0xc4, 0x5a, 0x91,
                0xb7, 0x77, 0xe0, 0xf7, 0xc8, 0x89,
            ]),
            chain_id: Some(1),
            name: "Trezor Steakhouse USDC Prime Vault".into(),
            asset_token: TokenInfo {
                symbol: "USDC".into(),
                decimals: 6,
                address: vec![
                    0xa0, 0xb8, 0x69, 0x91, 0xc6, 0x21, 0x8b, 0x36, 0xc1, 0xd1, 0x9d, 0x4a, 0x2e,
                    0x9e, 0xb0, 0xce, 0x36, 0x06, 0xeb, 0x48,
                ],
                chain_id: 1,
                name: "USD Coin".into(),
            },
            vault_token: TokenInfo {
                symbol: "trSHUSDCp".into(),
                decimals: 18,
                address: vec![
                    0xde, 0x6c, 0x23, 0xe5, 0x61, 0xf3, 0xe5, 0x58, 0x46, 0x20, 0x7e, 0xc4, 0x5a,
                    0x91, 0xb7, 0x77, 0xe0, 0xf7, 0xc8, 0x89,
                ],
                chain_id: 1,
                name: "Trezor Steakhouse USDC Prime Vault".into(),
            },
        },
        VaultInfo {
            // 0xE4DB1c5A1B709CE4d2adA6985D9D506e58F73829
            address: Some([
                0xe4, 0xdb, 0x1c, 0x5a, 0x1b, 0x70, 0x9c, 0xe4, 0xd2, 0xad, 0xa6, 0x98, 0x5d, 0x9d,
                0x50, 0x6e, 0x58, 0xf7, 0x38, 0x29,
            ]),
            chain_id: Some(1),
            name: "Trezor Steakhouse USDT Prime Vault".into(),
            asset_token: TokenInfo {
                symbol: "USDT".into(),
                decimals: 6,
                address: vec![
                    0xda, 0xc1, 0x7f, 0x95, 0x8d, 0x2e, 0xe5, 0x23, 0xa2, 0x20, 0x62, 0x06, 0x99,
                    0x45, 0x97, 0xc1, 0x3d, 0x83, 0x1e, 0xc7,
                ],
                chain_id: 1,
                name: "Tether USD".into(),
            },
            vault_token: TokenInfo {
                symbol: "trSHUSDTp".into(),
                decimals: 18,
                address: vec![
                    0xe4, 0xdb, 0x1c, 0x5a, 0x1b, 0x70, 0x9c, 0xe4, 0xd2, 0xad, 0xa6, 0x98, 0x5d,
                    0x9d, 0x50, 0x6e, 0x58, 0xf7, 0x38, 0x29,
                ],
                chain_id: 1,
                name: "Trezor Steakhouse USDT Prime Vault".into(),
            },
        },
    ]
}

pub(crate) fn get_token_label(token_addr: &[u8], network: &NetworkInfo) -> &'static str {
    // Only known token for now. We can use clear signing here to request token definitions from Connect, once it is done.
    // Would move it to EthereumVaultInfo but there don't seem to be plans to add more tokens for now.
    // https://etherscan.io/token/0x58d97b57bb95320f9a05dc918aef65434969c2b2
    const MORPHO_ADDR: [u8; 20] = [
        0x58, 0xd9, 0x7b, 0x57, 0xbb, 0x95, 0x32, 0x0f, 0x9a, 0x05, 0xdc, 0x91, 0x8a, 0xef, 0x65,
        0x43, 0x49, 0x69, 0xc2, 0xb2,
    ];

    if network.chain_id == 1 && token_addr == MORPHO_ADDR {
        "MORPHO"
    } else {
        "UNKNOWN"
    }
}

pub(crate) fn lookup_vault(network: &NetworkInfo, vault_addr: &[u8]) -> VaultInfo {
    for vault in known_vaults() {
        if let (Some(chain_id), Some(address)) = (vault.chain_id, vault.address) {
            if network.chain_id == chain_id && vault_addr == address {
                return vault;
            }
        }
    }
    unknown_vault()
}

File: core/src/apps/ethereum/yielding_vaults.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from buffer_types import AnyBytes
    from trezor.messages import EthereumNetworkInfo

    from .keychain import MsgInSignTx

from trezor.messages import EthereumTokenInfo

from .tokens import UNKNOWN_TOKEN

# Ethereum address bytes (20-byte)
_TEST_SH_USDC_VAULT_ADDRESS = (
    b"\xa5\x11\xd6\x18\xcd\x0f\x9d\x7c\xad\x79\x10\x09\xd7\xc5\xe3\xb1\x9c\x95\x68\xda"
)
_SH_USDC_VAULT_ADDRESS = (
    b"\xde\x6c\x23\xe5\x61\xf3\xe5\x58\x46\x20\x7e\xc4\x5a\x91\xb7\x77\xe0\xf7\xc8\x89"
)
_SH_USDT_VAULT_ADDRESS = (
    b"\xe4\xdb\x1c\x5a\x1b\x70\x9c\xe4\xd2\xad\xa6\x98\x5d\x9d\x50\x6e\x58\xf7\x38\x29"
)
_SH_ETH_VAULT_ADDRESS = (
    b"\x70\x4c\xfb\x08\x96\x90\x48\xa8\xdf\xf2\x98\xb2\x14\xf9\x59\x79\x1d\x8d\xa5\x09"
)
_USDC_ADDRESS = (
    b"\xa0\xb8\x69\x91\xc6\x21\x8b\x36\xc1\xd1\x9d\x4a\x2e\x9e\xb0\xce\x36\x06\xeb\x48"
)
_USDT_ADDRESS = (
    b"\xda\xc1\x7f\x95\x8d\x2e\xe5\x23\xa2\x20\x62\x06\x99\x45\x97\xc1\x3d\x83\x1e\xc7"
)
_WETH_ADDRESS = (
    b"\xc0\x2a\xaa\x39\xb2\x23\xfe\x8d\x0a\x0e\x5c\x4f\x27\xea\xd9\x08\x3c\x75\x6c\xc2"
)


class EthereumVaultInfo:
    def __init__(
        self,
        address: AnyBytes | None,
        chain_id: int | None,
        name: str,
        asset_token: EthereumTokenInfo,
        vault_token: EthereumTokenInfo,
    ) -> None:
        self.address = address
        self.chain_id = chain_id
        self.name = name
        self.asset_token = asset_token
        self.vault_token = vault_token


KNOWN_VAULTS = (
    # Test vault: https://etherscan.io/address/0xa511d618cD0F9d7cAD791009d7c5E3b19c9568da
    EthereumVaultInfo(
        address=_TEST_SH_USDC_VAULT_ADDRESS,
        chain_id=1,
        name="Test Steakhouse USDC Prime Vault",
        asset_token=EthereumTokenInfo(
            symbol="USDC",
            decimals=6,
            address=_USDC_ADDRESS,
            chain_id=1,
            name="USD Coin",
        ),
        vault_token=EthereumTokenInfo(
            symbol="tstSHUSDCp",
            decimals=18,
            address=_TEST_SH_USDC_VAULT_ADDRESS,
            chain_id=1,
            name="Test Steakhouse USDC Prime Vault",
        ),
    ),
    # https://etherscan.io/address/0xde6c23E561F3e55846207EC45A91b777e0F7C889
    EthereumVaultInfo(
        address=_SH_USDC_VAULT_ADDRESS,
        chain_id=1,
        name="Trezor Steakhouse USDC Prime Vault",
        asset_token=EthereumTokenInfo(
            symbol="USDC",
            decimals=6,
            address=_USDC_ADDRESS,
            chain_id=1,
            name="USD Coin",
        ),
        vault_token=EthereumTokenInfo(
            symbol="trSHUSDCp",
            decimals=18,
            address=_SH_USDC_VAULT_ADDRESS,
            chain_id=1,
            name="Trezor Steakhouse USDC Prime Vault",
        ),
    ),
    # https://etherscan.io/address/0xE4DB1c5A1B709CE4d2adA6985D9D506e58F73829
    EthereumVaultInfo(
        address=_SH_USDT_VAULT_ADDRESS,
        chain_id=1,
        name="Trezor Steakhouse USDT Prime Vault",
        asset_token=EthereumTokenInfo(
            symbol="USDT",
            decimals=6,
            address=_USDT_ADDRESS,
            chain_id=1,
            name="Tether USD",
        ),
        vault_token=EthereumTokenInfo(
            symbol="trSHUSDTp",
            decimals=18,
            address=_SH_USDT_VAULT_ADDRESS,
            chain_id=1,
            name="Trezor Steakhouse USDT Prime Vault",
        ),
    ),
    # https://etherscan.io/address/0x704cFb08969048a8DFf298B214F959791d8Da509
    EthereumVaultInfo(
        address=_SH_ETH_VAULT_ADDRESS,
        chain_id=1,
        name="Trezor Steakhouse ETH Prime Vault",
        asset_token=EthereumTokenInfo(
            symbol="WETH",
            decimals=18,
            address=_WETH_ADDRESS,
            chain_id=1,
            name="Wrapped Ether",
        ),
        vault_token=EthereumTokenInfo(
            symbol="trSHETHp",
            decimals=18,
            address=_SH_ETH_VAULT_ADDRESS,
            chain_id=1,
            name="Trezor Steakhouse ETH Prime Vault",
        ),
    ),
)

UNKNOWN_VAULT = EthereumVaultInfo(
    address=None,
    chain_id=None,
    name="UNKNOWN VAULT",
    asset_token=UNKNOWN_TOKEN,
    vault_token=UNKNOWN_TOKEN,
)


async def get_token_label(
    token_addr: AnyBytes,
    network: EthereumNetworkInfo,
    msg: MsgInSignTx,
    try_fetch_definitions: bool,
) -> str:
    # MORPHO is hardcoded. We support it regardless of Trezor Connect support.
    # https://etherscan.io/token/0x58d97b57bb95320f9a05dc918aef65434969c2b2
    _MORPHO_ADDR = b"\x58\xd9\x7b\x57\xbb\x95\x32\x0f\x9a\x05\xdc\x91\x8a\xef\x65\x43\x49\x69\xc2\xb2"
    if token_addr == _MORPHO_ADDR and network.chain_id == 1:
        return "MORPHO"

    if try_fetch_definitions and msg.supports_definition_request:
        from .clear_signing import request_definitions

        addr = bytes(token_addr)
        received_definitions, _ = await request_definitions(
            msg.chain_id, addr, func_sig=None
        )
        if received_definitions is not None:
            token = received_definitions.get_token(addr)
            if token is not UNKNOWN_TOKEN:
                return token.symbol

    return "UNKNOWN"


def lookup_vault(
    network: EthereumNetworkInfo, vault_addr: AnyBytes
) -> EthereumVaultInfo:
    """Returns the vault info for the given address and network"""

    for vault in KNOWN_VAULTS:
        if network.chain_id == vault.chain_id and vault_addr == vault.address:
            return vault

    return UNKNOWN_VAULT
